src/main/scala/cognite/spark/v1/DefaultSource.scala

```diff
@@ -188,104 +188,108 @@ class DefaultSource
-        }
-
-      dataRepartitioned.foreachPartition((rows: Iterator[Row]) => {
-        import CdpConnector.ioRuntime
 
-        val maxParallelism = config.parallelismPerPartition
-        val batches = Stream.fromIterator[IO](rows, chunkSize = batchSize).chunks
+    val relation: CdfRelation = resourceType match {
+      case AssetHierarchyBuilder.name =>
+        new AssetHierarchyBuilder(config)(sqlContext)
+      case NumericDataPointsRelation.name =>
+        new NumericDataPointsRelationV1(config)(sqlContext)
+      case StringDataPointsRelation.name =>
+        new StringDataPointsRelationV1(config)(sqlContext)
+      case EventsRelation.name =>
+        new EventsRelation(config)(sqlContext)
+      case TimeSeriesRelation.name =>
+        new TimeSeriesRelation(config)(sqlContext)
+      case AssetsRelation.name =>
+        new AssetsRelation(config)(sqlContext)
+      case FilesRelation.name =>
+        new FilesRelation(config)(sqlContext)
+      case SequenceRelation.name =>
+        new SequencesRelation(config)(sqlContext)
+      case LabelsRelation.name =>
+        new LabelsRelation(config)(sqlContext)
+      case SequenceRowsRelation.name =>
+        createSequenceRows(parameters, config, sqlContext)
+      case RelationshipsRelation.name =>
+        new RelationshipsRelation(config)(sqlContext)
+      case DataSetsRelation.name =>
+        new DataSetsRelation(config)(sqlContext)
+      case FlexibleDataModelBaseRelation.name =>
+        createFlexibleDataModelRelation(parameters, config, sqlContext)
+      case _ => sys.error(s"Resource type $resourceType does not support save()")
+    }
 
-        val operation: Seq[Row] => IO[Unit] = config.onConflict match {
-          case OnConflictOption.Abort =>
-            relation.insert
-          case OnConflictOption.Upsert =>
-            relation.upsert
-          case OnConflictOption.Update =>
-            relation.update
+    relation match {
+      case relation: DataPointsRelationV1[_] =>
+        if (config.onConflict == OnConflictOption.Delete) {
+          // Datapoints support 100_000 per request when inserting, but only 10_000 when deleting
+          val batchSize = config.batchSize.getOrElse(Constants.DefaultDataPointsLimit)
+          data.foreachPartition((rows: Iterator[Row]) => {
+            import CdpConnector.ioRuntime
+            val batches = rows.grouped(batchSize).toVector
+            batches.parTraverse_(relation.delete).unsafeRunSync()
+          })
+        } else {
+          // datapoints need special handling of dataframes and batches
+          relation.insert(data, overwrite = true)
+        }
+        relation
+      case relation: AssetHierarchyBuilder =>
+        config.onConflict match {
           case OnConflictOption.Delete =>
-            relation.delete
+            relation.delete(data)
+          case _ =>
+            relation.buildFromDf(data)
         }
+        relation
+      case relation: CdfRelation with WritableRelation =>
```

Comment:
hm, a little confusing naming from old code DefaultSource.createRelation is actually createRelationForWriting or something like it. But indeed it was checking against a list of relations that are supposed to support writing, now we check it via a trait so overall looks ok. By a quick look there were no writable sources not covered before so logic should be same as before :+1:

Reply:
This is also necessary to be able to call some stuff below (I believe the schema stuff notably)

Used to be this was in the "else" and then the typing of relation was enforced by the pattern matching, but in this case, we are pattern matching again on relation, which can be a AssetHierarchyBuilder which IIRC is not marked as writable, or something like that.

Basically some typing issue force this to be here, but it also makes perfect sense with the below code.